From e80c502633ca83fd3fc049d9d502a2ef28d39327 Mon Sep 17 00:00:00 2001
From: Peng Ma <peng.ma@nxp.com>
Date: Mon, 12 Nov 2018 11:13:01 +0000
Subject: [PATCH 719/767] dmaengine: fsl-dpaa2-qdma: add dest descriptor cmd
 write type for LX2160

LX2160 is different from other soc on dest descriptor cmd write type
so we should make distinctions

Signed-off-by: Peng Ma <peng.ma@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.12.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/dma/dpaa2-qdma/dpaa2-qdma.c | 22 +++++++++++++++++++---
 drivers/dma/dpaa2-qdma/dpaa2-qdma.h |  7 +++++++
 2 files changed, 26 insertions(+), 3 deletions(-)

diff --git a/drivers/dma/dpaa2-qdma/dpaa2-qdma.c b/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
index c444ccf010e8..82da2ddb76e0 100644
--- a/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
+++ b/drivers/dma/dpaa2-qdma/dpaa2-qdma.c
@@ -32,6 +32,7 @@
 #include <linux/types.h>
 #include <linux/delay.h>
 #include <linux/iommu.h>
+#include <linux/sys_soc.h>
 
 #include "../virt-dma.h"
 
@@ -141,7 +142,8 @@ static void dpaa2_qdma_populate_fd(uint32_t format,
 /* first frame list for descriptor buffer */
 static void dpaa2_qdma_populate_first_framel(
 		struct dpaa2_fl_entry *f_list,
-		struct dpaa2_qdma_comp *dpaa2_comp)
+		struct dpaa2_qdma_comp *dpaa2_comp,
+		bool wrt_changed)
 {
 	struct dpaa2_qdma_sd_d *sdd;
 
@@ -150,7 +152,12 @@ static void dpaa2_qdma_populate_first_framel(
 	/* source and destination descriptor */
 	sdd->cmd = cpu_to_le32(QDMA_SD_CMD_RDTTYPE_COHERENT); /* source descriptor CMD */
 	sdd++;
-	sdd->cmd = cpu_to_le32(QDMA_DD_CMD_WRTTYPE_COHERENT); /* dest descriptor CMD */
+
+	/* dest descriptor CMD */
+	if (wrt_changed)
+		sdd->cmd = cpu_to_le32(LX2160_QDMA_DD_CMD_WRTTYPE_COHERENT);
+	else
+		sdd->cmd = cpu_to_le32(QDMA_DD_CMD_WRTTYPE_COHERENT);
 
 	memset(f_list, 0, sizeof(struct dpaa2_fl_entry));
 	/* first frame list to source descriptor */
@@ -194,11 +201,15 @@ static struct dma_async_tx_descriptor *dpaa2_qdma_prep_memcpy(
 		dma_addr_t src, size_t len, unsigned long flags)
 {
 	struct dpaa2_qdma_chan *dpaa2_chan = to_dpaa2_qdma_chan(chan);
+	struct dpaa2_qdma_engine *dpaa2_qdma;
 	struct dpaa2_qdma_comp *dpaa2_comp;
 	struct dpaa2_fl_entry *f_list;
+	bool	wrt_changed;
 	uint32_t format;
 
+	dpaa2_qdma = dpaa2_chan->qdma;
 	dpaa2_comp = dpaa2_qdma_request_desc(dpaa2_chan);
+	wrt_changed = dpaa2_qdma->qdma_wrtype_fixup;
 
 #ifdef LONG_FORMAT
 	format = QDMA_FD_LONG_FORMAT;
@@ -212,7 +223,7 @@ static struct dma_async_tx_descriptor *dpaa2_qdma_prep_memcpy(
 
 #ifdef LONG_FORMAT
 	/* first frame list for descriptor buffer (logn format) */
-	dpaa2_qdma_populate_first_framel(f_list, dpaa2_comp);
+	dpaa2_qdma_populate_first_framel(f_list, dpaa2_comp, wrt_changed);
 
 	f_list++;
 #endif
@@ -839,6 +850,11 @@ static int dpaa2_qdma_probe(struct fsl_mc_device *dpdmai_dev)
 		goto err_reg;
 	}
 
+	if (soc_device_match(soc_fixup_tuning))
+		dpaa2_qdma->qdma_wrtype_fixup = true;
+	else
+		dpaa2_qdma->qdma_wrtype_fixup = false;
+
 	dma_cap_set(DMA_PRIVATE, dpaa2_qdma->dma_dev.cap_mask);
 	dma_cap_set(DMA_SLAVE, dpaa2_qdma->dma_dev.cap_mask);
 	dma_cap_set(DMA_MEMCPY, dpaa2_qdma->dma_dev.cap_mask);
diff --git a/drivers/dma/dpaa2-qdma/dpaa2-qdma.h b/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
index ab682a41e006..7a590b2490e8 100644
--- a/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
+++ b/drivers/dma/dpaa2-qdma/dpaa2-qdma.h
@@ -67,6 +67,7 @@ struct dpaa2_qdma_sd_d {
 /* Destination descriptor command write transaction type for RBP=0:
  coherent copy of cacheable memory */
 #define QDMA_DD_CMD_WRTTYPE_COHERENT (0x6 << 28)
+#define LX2160_QDMA_DD_CMD_WRTTYPE_COHERENT (0xb << 28)
 
 #define QDMA_SG_FMT_SDB		0x0 /* single data buffer */
 #define QDMA_SG_FMT_FDS		0x1 /* frame data section */
@@ -180,6 +181,7 @@ struct dpaa2_qdma_engine {
 	struct dma_device	dma_dev;
 	u32			n_chans;
 	struct dpaa2_qdma_chan	chans[NUM_CH];
+	bool			qdma_wrtype_fixup;
 
 	struct dpaa2_qdma_priv *priv;
 };
@@ -216,6 +218,11 @@ struct dpaa2_qdma_priv_per_prio {
 	struct dpaa2_qdma_priv *priv;
 };
 
+static struct soc_device_attribute soc_fixup_tuning[] = {
+	{ .family = "QorIQ LX2160A"},
+	{ },
+};
+
 /* FD pool size: one FD + 3 Frame list + 2 source/destination descriptor */
 #define FD_POOL_SIZE (sizeof(struct dpaa2_fd) + \
 		sizeof(struct dpaa2_fl_entry) * 3 + \
-- 
2.17.0

